AWS Metadata Processor Plugins

johnrengelman opened the pull request:
### Required for all PRs:

- [x] CHANGELOG.md updated (we recommend not updating this until the PR has been approved by a maintainer)
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)
- [x] README.md updated (if adding a new plugin)

johnrengelman commented:
Having a bit of a crisis of design on this. This PR implements it as a separate processor for each AWS service, but I'm wondering if there should just be a single AWS processor that you can enable/configure each service as sub-sections for it.

johnrengelman commented:
@sparrc any thoughts on design for this?

sparrc commented:
sorry @johnrengelman, but I haven't had time to review this. As it's a plugin request it likely will be a candidate for moving into our external plugins repository, of which we're trying to get a prototype going for 1.2.

My preference would probably be to keep them separate.

You'll also need to rebase the changes as we've made some significant changes to the telegraf Metric implementation. AddTag is now available so you can revert that portion of your change.

johnrengelman commented:
@sparrc I've rebased this onto the current `master` and added a couple more services. I think this would be a good start to get in and provide a framework for future AWS services/properties.

johnrengelman commented:
This is going to need some caching I believe.
Also if there's more info on the external plugins deal, I'd be happy to start working on that.

johnrengelman commented:
@sparrc This is ready to be looked at unless you want to point me to the external plugin stuff so I can start looking at implementing it that way. 
I've been running this code in production for a month now without issues.

danielnelson commented on plugins/processors/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,213 @@
+
+  ## Specify the metric names to annotate with EC2 metadata
+  ## By default is configured for "cloudwatch_aws_ec2", the default output from the Cloudwatch input plugin
+  #metric_names = [ "cloudwatch_aws_ec2" ]
```
Metrics can be selected using namepass/namedrop/tagpass/tagdrop, so I think we can remove this.

danielnelson commented on plugins/processors/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,213 @@
+}
+
+func (e *EC2) getInstanceForMetric(metric telegraf.Metric) (*ec2.Instance, error) {
+	id := metric.Tags()[e.Id]
```
Use `metric.GetTag(`, added long after you wrote this pull request, because it doesn't allocate.

danielnelson commented on plugins/processors/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,213 @@
+	if e.fetched.IsZero() {
+		e.fetched = time.Now()
+	}
+	if time.Since(e.fetched) >= e.ttl {
```
It would be nice to have a way to clean out old instances from the cache.  Perhaps if the length of the map is larger than some value?

danielnelson commented on plugins/processors/aws/ec2/ec2.go:
```diff
@@ -0,0 +1,213 @@
+  #instance_type = true
+
+  ## Enable annotating metrics with the AMI ID
+  #ami_id = true
```
I would probably make these a list of strings.  This would be optional though, as most of the problems with boolean options don't really apply here, namely that they often end up being interdependent and they don't scale up well.

danielnelson commented:
@johnrengelman Sorry it has taken so long, I really like these plugins and think they would be great examples for "enriching" data.  I only reviewed the ec2 plugin, but if you are still interested I can review the others.  If possible could you clear up the conflicts and look at my comments around the ec2 plugin?

glinton commented:
@johnrengelman Are you still interested/available to continue this?
